MMI + MBR training

csukuangfj opened the pull request:
~working in progress.~

---

Ready for review.

csukuangfj commented on egs/librispeech/asr/simple_v1/mmi_mbr_train.py:
```diff
@@ -0,0 +1,515 @@
+                                       seqframe_idx_name='seqframe_idx')
+    logging.info('mbr_den_graph {}'.format(mbr_den.shape))
+
+    mbr_den_graph = k2.intersect_dense_pruned(mbr_den,
```
The training gets stuck here. `k2.intersect_dense_prune` seems never terminating.
I have decreased search beam and max active states.

danpovey replied:
mm I may need to debug this.

csukuangfj commented on snowfall/training/mmi_mbr_graph.py:
```diff
@@ -0,0 +1,170 @@
+
+class MmiMbrTrainingGraphCompiler(object):
+
+    def __init__(self,
```
This is the script for building graphs. I am not sure if it is correct.

danpovey commented:
Getting this warning when running it:
```
Last login: Wed Jan 13 11:23:15 on ttys013

The default interactive shell is now zsh.
To update your account to use zsh, please run `chsh -s /bin/zsh`.
For more details, please visit https://support.apple.com/kb/HT208050.
mac:matrix: 








































mac:matrix: ls
Makefile			kaldi-blas.h			kaldi-vector.o			matrix-lib-test.o		packed-matrix.o			sparse-matrix.cc
Matrix.vcxproj			kaldi-matrix-inl.h		matrix-common.h			matrix-lib.h			qr.cc				sparse-matrix.h
cblas-wrappers.h		kaldi-matrix.a			matrix-functions-inl.h		numpy-array-test.cc		qr.o				sparse-matrix.o
compressed-matrix.cc		kaldi-matrix.cc			matrix-functions.cc		numpy-array.cc			sp-matrix-inl.h			srfft.cc
compressed-matrix.h		kaldi-matrix.h			matrix-functions.h		numpy-array.h			sp-matrix.cc			srfft.h
compressed-matrix.o		kaldi-matrix.h~			matrix-functions.o		optimization.cc			sp-matrix.h			srfft.o
foo				kaldi-matrix.o			matrix-lib-speed-test.cc	optimization.h			sp-matrix.o			test_data
hEulksI636RvUfO0?mu=1		kaldi-vector-inl.h		matrix-lib-test			optimization.o			sparse-matrix-test		tp-matrix.cc
jama-eig.h			kaldi-vector.cc			matrix-lib-test.cc		packed-matrix.cc		sparse-matrix-test.cc		tp-matrix.h
jama-svd.h			kaldi-vector.h			matrix-lib-test.cc~		packed-matrix.h			sparse-matrix-test.o		tp-matrix.o
mac:matrix: ssh relay.xiaomi.com
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
请[填写密码]或使用小米人[扫描二维码]进行登录。如看不到二维码，请把终端模拟软件的颜色支持打开。
FAQ: http://wiki.n.miui.com/pages/viewpage.action?pageId=30213224

Your [EMAIL] password: login via qrcode
                                       Last login: Wed Jan 13 11:13:36 2021 from 123.116.244.178
                             oo$$$$$$$$$$oo
                          oooo$$$$$$$$$$$$oooo
                      oo$$$$$$$$$$$$$$$$$$$$$$$$o
                   oo$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$o         o$   $$ o$
   o $ oo        o$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$o       $$ $$ $$o$
oo $ $ "$      o$$$$$$$$$    $$$$$$$$$$$$$    $$$$$$$$$o       $$$o$$o$
"$$$$$$o$     o$$$$$$$$$      $$$$$$$$$$$      $$$$$$$$$$o    $$$$$$$$
  $$$$$$$    $$$$$$$$$$$      $$$$$$$$$$$      $$$$$$$$$$$$$$$$$$$$$$$
  $$$$$$$$$$$$$$$$$$$$$$$    $$$$$$$$$$$$$    $$$$$$$$$$$$$$  """$$$
   "$$$""""$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$     "$$$
    $$$   o$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$     "$$$o
   o$$"   $$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$       $$$o
   $$$    $$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$" "$$$$$$ooooo$$$$o
  o$$$oooo$$$$$  $$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$   o$$$$$$$$$$$$$$$$$
  $$$$$$$$"$$$$   $$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$     $$$$""""""""
 """"       $$$$    "$$$$$$$$$$$$$$$$$$$$$$$$$$$$"      o$$$
            "$$$o     """$$$$$$$$$$$$$$$$$$"$$"         $$$
              $$$o          "$$""$$$$$$""""           o$$$
               $$$$o                                o$$$"
                "$$$$o      o$$$$$$o"$$$$o        o$$$$
                  "$$$$$oo     ""$$$$o$$$$$o   o$$$$""
                     ""$$$$$oooo  "$$$o$$$$$$$$$"""
                        ""$$$$$$$oo $$$$$$$$$$
                                """"$$$$$$$$$$$
                                    $$$$$$$$$$$$
                                     $$$$$$$$$$"
                                      "$$$""""
                                      "$$$""""

【用法】
* 直接在relay-shell>提示符后面输入主机名，然后按回车即可登录：
    relay-shell> c3-pt-relay02.bj <按回车>
* 或者输入shell进入bash，以进行批量hd登录等操作
    relay-shell> shell <按回车>
* 在任何情况下连按两次tab键可以自动补全
[Usage]
Type ip address or hostname to connect to the speificed host, 
or commands listed below to launch the corresponding feature.
* Use TAB for auto-complete, CTRL+R for reverse-search
* Try pressing the Tab key twice to see all the hosts you have access to

Command list:
	shell:		where you show off your bash programming skills
	help:		show this message
	exit		exit this place

bj5-pt-relay01.alicn.idc.xiaomi.com:dpovey> shell
Launching an interactive shell
不要在本机磁盘存储重要数据，常用脚本请做好备份。
dpovey@bj5-pt-relay01: ~ $ ssh dpovey@10.177.36.72
dpovey@10.177.36.72's password: 
Last login: Thu Jan 14 21:22:31 2021 from 10.53.13.8
de-74279-k2-dev-0929153814-7686b86bc5-qbfjq:~: tmux attach

[1]+  Stopped                 emacs -nw patent_fsa.txt
mac:notes: %
emacs -nw patent_fsa.txt

[1]+  Stopped                 emacs -nw patent_fsa.txt
mac:notes: pwd
/Users/dpovey/notes
mac:notes: ls -l ~/kaldi/egs/tedlium/s5/utils
lrwxr-xr-x  1 dpovey  staff  18 Jul  2  2019 /Users/dpovey/kaldi/egs/tedlium/s5/utils -> ../../wsj/s5/utils
mac:notes: %
emacs -nw patent_fsa.txt

File Edit Options Buffers Tools C++ Help                                                                                                                                                           

  bool continuous_;

  const DropoutMaskComponent &operator
  = (const DropoutMaskComponent &other); // Disallow.                                                                                                                                              
};



/**                                                                                                                                                                                                
   GeneralDropoutComponent implements dropout, including a continuous                                                                                                                              
   variant where the thing we multiply is not just zero or one, but may                                                                                                                            
   be a continuous value.  It is intended for the case where you want to                                                                                                                           
   either share the dropout mask across all of time, or across groups                                                                                                                              
   of 't' values (e.g. the first block of 10 values gets one dropout                                                                                                                               
   mask, the second block of 10 gets another one, and so on).                                                                                                                                      
                                                                                                                                                                                                   
   It also has support for the frequency component of SpecAugment.                                                                                                                                 
                                                                                                                                                                                                   
   Configuration values accepted on the command line, with defaults:                                                                                                                               
-UU-:----F1  nnet-general-component.h   73% L804  Git-master  (C++//l Abbrev) ---------------------------------------------------------------------------------------------------------------------

GeneralDropoutComponent::GeneralDropoutComponent():
    dim_(-1), block_dim_(-1), time_period_(0),
    dropout_proportion_(0.5),
    specaugment_max_proportion_(0.0),
    specaugment_max_regions_(1),
    continuous_(false) { }

GeneralDropoutComponent::GeneralDropoutComponent(
    const GeneralDropoutComponent &other):
    RandomComponent(other),
    dim_(other.dim_),
    block_dim_(other.block_dim_),
    time_period_(other.time_period_),
    dropout_proportion_(other.dropout_proportion_),
    specaugment_max_proportion_(other.specaugment_max_proportion_),
    specaugment_max_regions_(other.specaugment_max_regions_),
    continuous_(other.continuous_) { }

-UU-:----F1  nnet-general-component.cc   74% L1554 Git:master  (C++//l Abbrev) --------------------------------------------------------------------------------------------------------------------

  [Restored Jan 20, 2021 at 5:51:38 PM]
Last login: Wed Jan 20 17:51:38 on ttys011

The default interactive shell is now zsh.
To update your account to use zsh, please run `chsh -s /bin/zsh`.
For more details, please visit https://support.apple.com/kb/HT208050.
mac:notes: 
mac:notes: pwd
/Users/dpovey/notes
mac:notes: 








































mac:notes: jobs
mac:notes: jobs
mac:notes: pwd
/Users/dpovey/notes
mac:notes: ls -l ht | head
ls: ht: No such file or directory
mac:notes: ls -lht | head
total 5984
-rw-r--r--  1 dpovey  staff   6.9K Jan 20 16:08 patent_fsa.txt
-rw-r--r--  1 dpovey  staff   4.2K Jan 19 17:43 2020.highlight
-rw-r--r--  1 dpovey  staff   315B Jan 19 14:29 patent_fsa.txt~
-rw-r--r--  1 dpovey  staff   2.9K Jan 14 15:41 2020.highlight~
-rw-r--r--  1 dpovey  staff   281B Jan  3 20:14 jan3_2021.txt
-rw-r--r--  1 dpovey  staff   259B Jan  3 20:13 jan3_2021.txt~
drwxr-xr-x@ 2 dpovey  staff    64B Jan  3 20:12 updates
-rw-r--r--  1 dpovey  staff   3.1K Dec 14 18:36 k2_intro
-rw-r--r--  1 dpovey  staff   239B Dec 14 16:12 k2_intro~
mac:notes: emacs -nw patent_fsa.txt

[1]+  Stopped                 emacs -nw patent_fsa.txt
mac:notes: bg
[1]+ emacs -nw patent_fsa.txt &

[1]+  Stopped                 emacs -nw patent_fsa.txt
mac:notes: %
emacs -nw patent_fsa.txt

















File Edit Options Buffers Tools Text Help                                                                                                                                                            

 - In the next 2-3 weeks, plan to finish and test MBR training and start
   writing code for n-best list rescoring.

===


Jan 3rd:
 Progress this week:
    Currently debugging ASG training
    Librispeech WER with ASG training is 12% (previously we had
      15% with CTC).
      Note: these numbers are still not very competitive, we are still fixing
      various things and have not really tuned our system.

===========


======

  Next-gen Kaldi


 What is Kaldi

   Kaldi is an open-source toolkit for automatic speech recognition (ASR) that
   has existed since 2011; it is the most popular speech recognition toolkit.
   Since 2011 it has evolved considerably as the state of the art has
   developed: for instance, adding support for neural network acoustic models.
   For most of that time the chief maintainer of Kaldi (Dan Povey) has worked
   at Johns Hopkins University in the USA, but in December 2019 he joined
   Xiaomi in Beijing.

 The project

  Kaldi's neural network support is provided by custom neural network
  code that is not as flexible as toolkits like PyTorch and TensorFlow.
  For a long time there has been a demand for Kaldi to support external neural
  network toolkits.  The aim of the "next-gen Kaldi" projectd is to add such
  support.

 The team
-UU-:----F1  progress.txt   16% L17    (Text) -------------------------------------------------------------------------------------------------------------------------------------------------------

  [Restored Jan 26, 2021 at 6:21:31 PM]
Last login: Tue Jan 26 18:21:30 on ttys012

The default interactive shell is now zsh.
To update your account to use zsh, please run `chsh -s /bin/zsh`.
For more details, please visit https://support.apple.com/kb/HT208050.
mac:notes: 
mac:notes: ssh relay.xiaomi.com
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
                                                    
请[填写密码]或使用小米人[扫描二维码]进行登录。如看不到二维码，请把终端模拟软件的颜色支持打开。
FAQ: http://wiki.n.miui.com/pages/viewpage.action?pageId=30213224

Your [EMAIL] password: login via qrcode
                                       Last login: Tue Jan 26 22:28:14 2021 from 123.116.246.225
                             oo$$$$$$$$$$oo
                          oooo$$$$$$$$$$$$oooo
                      oo$$$$$$$$$$$$$$$$$$$$$$$$o
                   oo$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$o         o$   $$ o$
   o $ oo        o$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$o       $$ $$ $$o$
oo $ $ "$      o$$$$$$$$$    $$$$$$$$$$$$$    $$$$$$$$$o       $$$o$$o$
"$$$$$$o$     o$$$$$$$$$      $$$$$$$$$$$      $$$$$$$$$$o    $$$$$$$$
  $$$$$$$    $$$$$$$$$$$      $$$$$$$$$$$      $$$$$$$$$$$$$$$$$$$$$$$
  $$$$$$$$$$$$$$$$$$$$$$$    $$$$$$$$$$$$$    $$$$$$$$$$$$$$  """$$$
   "$$$""""$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$     "$$$
    $$$   o$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$     "$$$o
   o$$"   $$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$       $$$o
   $$$    $$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$" "$$$$$$ooooo$$$$o
  o$$$oooo$$$$$  $$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$   o$$$$$$$$$$$$$$$$$
  $$$$$$$$"$$$$   $$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$$     $$$$""""""""
 """"       $$$$    "$$$$$$$$$$$$$$$$$$$$$$$$$$$$"      o$$$
            "$$$o     """$$$$$$$$$$$$$$$$$$"$$"         $$$
              $$$o          "$$""$$$$$$""""           o$$$
               $$$$o                                o$$$"
                "$$$$o      o$$$$$$o"$$$$o        o$$$$
                  "$$$$$oo     ""$$$$o$$$$$o   o$$$$""
                     ""$$$$$oooo  "$$$o$$$$$$$$$"""
                        ""$$$$$$$oo $$$$$$$$$$
                                """"$$$$$$$$$$$
                                    $$$$$$$$$$$$
                                     $$$$$$$$$$"
                                      "$$$""""
                                      "$$$""""

【用法】
* 直接在relay-shell>提示符后面输入主机名，然后按回车即可登录：
    relay-shell> c3-pt-relay02.bj <按回车>
* 或者输入shell进入bash，以进行批量hd登录等操作
    relay-shell> shell <按回车>
* 在任何情况下连按两次tab键可以自动补全
[Usage]
Type ip address or hostname to connect to the speificed host, 
or commands listed below to launch the corresponding feature.
* Use TAB for auto-complete, CTRL+R for reverse-search
* Try pressing the Tab key twice to see all the hosts you have access to

Command list:
	shell:		where you show off your bash programming skills
	help:		show this message
	exit		exit this place

bj6-pt-relay01.alicn.idc.xiaomi.com:dpovey> shell
Launching an interactive shell
不要在本机磁盘存储重要数据，常用脚本请做好备份。
dpovey@bj6-pt-relay01: ~ $ ssh dpovey@10.177.36.72
dpovey@10.177.36.72's password: 
Last login: Tue Jan 26 22:28:34 2021 from 10.53.13.8
de-74279-k2-dev-0929153814-7686b86bc5-qbfjq:~: tmux attach

File Edit Options Buffers Tools Complete In/Out Signals Help                                                                                                                                       
  File "./mmi_mbr_train.py", line 462, in main
    objf = train_one_epoch(dataloader=train_dl,
  File "./mmi_mbr_train.py", line 257, in train_one_epoch
    get_objf(batch, model, P, device, graph_compiler, True, optimizer)
  File "./mmi_mbr_train.py", line 109, in get_objf
    num, den, mbr_num, mbr_den = graph_compiler.compile(texts, P)
  File "/ceph-dan/.local/lib/python3.8/site-packages/snowfall-0.1-py3.8.egg/snowfall/training/mmi_mbr_graph.py", line 125, in compile
    logging.info('compiling', len(texts))
Message: 'compiling'
Arguments: (23,)
--- Logging error ---
Traceback (most recent call last):
  File "/usr/lib/python3.8/logging/__init__.py", line 1081, in emit
    msg = self.format(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 925, in format
    return fmt.format(record)
  File "/usr/lib/python3.8/logging/__init__.py", line 664, in format
    record.message = record.getMessage()
  File "/usr/lib/python3.8/logging/__init__.py", line 369, in getMessage
    msg = msg % self.args
TypeError: not all arguments converted during string formatting
Call stack:
  File "./mmi_mbr_train.py", line 515, in <module>
    main()
  File "./mmi_mbr_train.py", line 462, in main
    objf = train_one_epoch(dataloader=train_dl,
  File "./mmi_mbr_train.py", line 257, in train_one_epoch
    get_objf(batch, model, P, device, graph_compiler, True, optimizer)
  File "./mmi_mbr_train.py", line 109, in get_objf
    num, den, mbr_num, mbr_den = graph_compiler.compile(texts, P)
  File "/ceph-dan/.local/lib/python3.8/site-packages/snowfall-0.1-py3.8.egg/snowfall/training/mmi_mbr_graph.py", line 125, in compile
    logging.info('compiling', len(texts))
Message: 'compiling'
..
```

csukuangfj commented:
It is not an error. I am fixing it and remove the extra graph.

danpovey commented:
OK.  I get this error later:
```
Arguments: (23,)
2021-01-28 22:28:41,373 INFO [mmi_mbr_graph.py:133] mbr num
2021-01-28 22:28:41,414 INFO [mmi_mbr_graph.py:138] num
2021-01-28 22:28:41,453 INFO [mmi_mbr_graph.py:143] den
2021-01-28 22:28:41,495 INFO [mmi_mbr_graph.py:146] mbr den
Traceback (most recent call last):
  File "./mmi_mbr_train.py", line 515, in <module>
    main()
  File "./mmi_mbr_train.py", line 462, in main
    objf = train_one_epoch(dataloader=train_dl,
  File "./mmi_mbr_train.py", line 257, in train_one_epoch
    get_objf(batch, model, P, device, graph_compiler, True, optimizer)
  File "./mmi_mbr_train.py", line 109, in get_objf
    num, den, mbr_num, mbr_den = graph_compiler.compile(texts, P)
  File "/ceph-dan/.local/lib/python3.8/site-packages/snowfall-0.1-py3.8.egg/snowfall/training/mmi_mbr_graph.py", line 147, in compile
  File "/ceph-dan/.local/lib/python3.8/site-packages/k2/utils.py", line 231, in create_fsa_vec
    value = torch.cat(values)
TypeError: expected Tensor as element 0 in argument 0, but got _k2.RaggedInt
```

csukuangfj commented:
> OK. I get this error later:

k2.determinize(L, G) results in an FSA with ragged int as aux_labels.

Please use the latest k2.  I have fixed k2.create_fsa_vec() for FSAs with ragged int tensor as attributes.
Fixed in https://github.com/k2-fsa/k2/pull/634

csukuangfj commented:
I have removed the extra graph. Now there are only 3 graphs:

- mmi_num
- mmi_den


The above two graphs are from mmi training.

The third graph is `decoding_graph = compose(ctc_topo, L, G)`

The loss consists of two part:
- The one from MMI training
- The sum of abs difference of two sparse matrices:
  - The first sparse matrix is from mmi_num (using `intersect_dense`)
  - The second is from decoding_graph (using `intersect_dense_pruned`, however, it never terminates)

csukuangfj commented:
@pzelasko 
Thanks! Will look into it.

csukuangfj commented on egs/librispeech/asr/simple_v1/mmi_mbr_train.py:
```diff
@@ -0,0 +1,508 @@
+    nnet_output = nnet_output.permute(0, 2, 1)  # now nnet_output is [N, T, C]
+
+    if is_training:
+        num, den, decoding_graph = graph_compiler.compile(texts, P)
```
`decoding_graph` is added for MBR training.

csukuangfj commented on egs/librispeech/asr/simple_v1/mmi_mbr_train.py:
```diff
@@ -0,0 +1,508 @@
+
+    mmi_loss = -tot_score
+
+    total_loss = mmi_loss + mbr_loss
```
Here is the total loss.

danpovey replied:
OK..

danpovey replied:
Would be nice if you can verify that if you make this `total_loss = mmi_loss + 0.5 * mbr_loss`, the MMI objf gets better and the MBR objf gets worse (in the plots).

csukuangfj replied:
Will verify it before merging.

csukuangfj replied:
> the MMI objf gets better and the MBR objf gets worse (in the plots).

Both losses are decreasing. MBR losses do not get worse.
Please see https://tensorboard.dev/experiment/QFeH3h7qTgOePVhWoxKi8Q/#scalars&_smoothingWeight=0

Does this reflect any problems?

danpovey replied:
I just mean I'd expect that the MBR loss would not reach quite as good as value as when the weight is 1.0, and the MMI loss would reach a slightly better value.

csukuangfj replied:
I see. Will report the final WER after 10 epochs tomorrow.

csukuangfj commented:
Here is some output from gdb using

- `gdb -ex r --args python ./mmi_mbr_train.py`
- ctrl + C, bt, continue, ctr+C, bt, continue (repeat for several times)

Seems it is stuck in `k2::MultiGraphDenseIntersectPruned::PropagateForward`. `t` is always 1.

```
#12 0x00007ffd867abd22 in k2::Array1<int>::operator[] (this=0x7fffffffc760, i=109819) 
at /root/fangjun/open-source/k2/k2/csrc/array.h:272

#13 0x00007ffd867a96a1 in k2::Array1<int>::Back (this=0x7fffffffc760) 
at /root/fangjun/open-source/k2/k2/csrc/array.h:284

#14 0x00007ffd867a4775 in k2::Renumbering::ComputeOld2New (this=0x7fffffffc740) 
at /root/fangjun/open-source/k2/k2/csrc/algorithms.cu:21

#15 0x00007ffd86919540 in k2::Renumbering::NumNewElems (this=0x7fffffffc740) 
at /root/fangjun/open-source/k2/k2/csrc/algorithms.h:81

#16 0x00007ffd8692590a in 
k2::MultiGraphDenseIntersectPruned::PropagateForward<32> (this=0x7fffffffca50, t=1, cur_frame=0x5555589a60e0)
    at /root/fangjun/open-source/k2/k2/csrc/intersect_dense_pruned.cu:754

#17 0x00007ffd8691aa44 in k2::MultiGraphDenseIntersectPruned::Intersect (this=0x7fffffffca50)
    at /root/fangjun/open-source/k2/k2/csrc/intersect_dense_pruned.cu:257
```

danpovey commented:
After updating to latest code, still have this error:
```
================================================================================
2021-01-28 23:06:34,057 INFO [mmi_mbr_train.py:453] epoch 0, learning rate 0.001
2021-01-28 23:06:34,802 INFO [mmi_mbr_graph.py:127] compiling
2021-01-28 23:06:35,262 INFO [mmi_mbr_graph.py:135] num
2021-01-28 23:06:35,301 INFO [mmi_mbr_graph.py:140] den
Traceback (most recent call last):
  File "./mmi_mbr_train.py", line 508, in <module>
    main()
  File "./mmi_mbr_train.py", line 455, in main
    objf = train_one_epoch(dataloader=train_dl,
  File "./mmi_mbr_train.py", line 250, in train_one_epoch
    get_objf(batch, model, P, device, graph_compiler, True, optimizer)
  File "./mmi_mbr_train.py", line 109, in get_objf
    num, den, decoding_graph = graph_compiler.compile(texts, P)
  File "/ceph-dan/.local/lib/python3.8/site-packages/snowfall-0.1-py3.8.egg/snowfall/training/mmi_mbr_graph.py", line 143, in compile
  File "/ceph-dan/.local/lib/python3.8/site-packages/k2/utils.py", line 231, in create_fsa_vec
    value = torch.cat(values)
TypeError: expected Tensor as element 0 in argument 0, but got _k2.RaggedInt
```

csukuangfj commented:
What is the output of
```
$ python3 -m k2.version
```
?

It will show the git SHA-1, which should refer to the latest k2

danpovey commented:
Old version-- odd.  Maybe my install script is not working..  Will look into it.
```
k2 version: 0.1.3
Build type: Debug
Git SHA1: 40323a2a7f93222619386ff9d748cdf2095703d7
Git date: Wed Jan 13 11:16:26 2021
Cuda used to build k2: 10.1
cuDNN used to build k2: 7.6.2
Python version used to build k2: 3.8
OS used to build k2: Ubuntu 18.04.5 LTS
CMake version: 3.10.2
GCC version: 7.5.0
CMAKE_CUDA_FLAGS:  -D_GLIBCXX_USE_CXX11_ABI=0  --expt-extended-lambda -gencode arch=compute_35,code=sm_35 --expt-extended-lambda -gencode arch=compute_50,code=sm_50 --expt-extended-lambda -genco\
de arch=compute_60,code=sm_60 --expt-extended-lambda -gencode arch=compute_61,code=sm_61 --expt-extended-lambda -gencode arch=compute_70,code=sm_70 --expt-extended-lambda -gencode arch=compute_7\
5,code=sm_75 --compiler-options -Wall --compiler-options -Wno-unknown-pragmas
CMAKE_CXX_FLAGS:  -D_GLIBCXX_USE_CXX11_ABI=0
PyTorch version used to build k2: 1.7.0+cu101
PyTorch is using Cuda: 10.1
NVTX enabled: True
Disable debug: False
Sync kernels : False
Disable checks: False
```

csukuangfj commented:
The following script may be helpful:

```bash
export PYTHONPATH=/root/fangjun/open-source/k2/k2/python:$PYTHONPATH
export PYTHONPATH=/root/fangjun/open-source/k2/build/lib:$PYTHONPATH
export PYTHONPATH=/root/fangjun/open-source/snowfall:$PYTHONPATH
export PYTHONPATH=/root/fangjun/open-source/lhotse:$PYTHONPATH
```

No need to install k2, snowfall, or lhotse.

To use the latest k2, just compile it. To verify that snowfall is using the one you just compiled, use `python3 -m k2.version`
and compare the `git SHA-1` from the output with the one in your k2 repo.

csukuangfj commented:
Also,

```bash
python3 -c "import k2; print(k2.__file__)"
```
should point to the file in the k2 source tree.

danpovey commented:
Mm.  If something goes wrong in the hash-code kernels, they can loop forever.  Likely that's what happened, 
e.g. Insert() at intersect_dense_pruned.cu:746 is looping.  It's a generic error, unfortunately, e.g. could be due to
too-full hash...  I'm looking into it.

csukuangfj commented:
>  I'm looking into it.

Thanks

danpovey commented:
OK I fixed the hash thing.  New error:
```
[Thread 0x7fff14110700 (LWP 3629510) exited]
Traceback (most recent call last):
  File "./mmi_mbr_train.py", line 508, in <module>
    main()
  File "./mmi_mbr_train.py", line 455, in main
    objf = train_one_epoch(dataloader=train_dl,
  File "./mmi_mbr_train.py", line 250, in train_one_epoch
    get_objf(batch, model, P, device, graph_compiler, True, optimizer)
  File "./mmi_mbr_train.py", line 161, in get_objf
    cols=num_graph.phones,
  File "/ceph-dan/.local/lib/python3.8/site-packages/k2/fsa.py", line 407, in __getattr__
    raise AttributeError(f'Unknown attribute {name}')
AttributeError: Unknown attribute phones
[Thread 0x7ffeea03d700 (LWP 3629541) exited]
[Thread 0x7ffeea88f700 (LWP 3629540) exited]
[Thread 0x7fff118be700 (LWP 3629501) exited]
[Thread 0x7fff0f06c700 (LWP 3629500) exited]
[Thread 0x7fff0c81a700 (LWP 3629484) exited]
[Inferior 1 (process 3629409) exited with code 01]
(gdb) 

```

csukuangfj commented:
> OK I fixed the hash thing. New error:


Fixed, please pull the latest code of this pull-request.

csukuangfj commented:
It is able to get the loss now.

danpovey commented:
```
de-74279-k2-dev-0929153814-7686b86bc5-qbfjq:simple_v1: python3 mmi_mbr_train.py 
Traceback (most recent call last):
  File "mmi_mbr_train.py", line 16, in <module>
    import k2.sparse
ModuleNotFoundError: No module named 'k2.sparse'
```

danpovey commented:
trying to redo `cmake ..` and rebuild k2, maybe build problem..

csukuangfj commented:
> ModuleNotFoundError: No module named 'k2.sparse'


Please use the latest code of this pull-request. It has been fixed.

danpovey commented:
will go to bed now.  Great work!

csukuangfj commented:
Thanks!!

Here is the training log. The loss is decreasing.  It will print the WER after training and decoding (by running run.sh).

We will see.

Going to bed too.

```
2021-01-29 00:08:15,527 INFO [mmi_mbr_train.py:443] epoch 0, learning rate 0.001
2021-01-29 00:08:21,741 INFO [mmi_mbr_train.py:247] batch 0, epoch 0/10 global average objf: 3.146235 over 9623.0 frames (100.0% kept), current batch
average objf: 3.146234 over 9623 frames (100.0% kept) avg time waiting for batch 0.529s
2021-01-29 00:09:16,275 INFO [mmi_mbr_train.py:247] batch 10, epoch 0/10 global average objf: 2.766029 over 107435.0 frames (100.0% kept), current batch average objf: 2.482606 over 9835 frames (100.0% kept) avg time waiting for batch 0.056s
2021-01-29 00:10:09,405 INFO [mmi_mbr_train.py:247] batch 20, epoch 0/10 global average objf: 2.529759 over 205210.0 frames (100.0% kept), current batch average objf: 2.127868 over 9856 frames (100.0% kept) avg time waiting for batch 0.030s
2021-01-29 00:11:03,684 INFO [mmi_mbr_train.py:247] batch 30, epoch 0/10 global average objf: 2.407867 over 302507.0 frames (100.0% kept), current batch average objf: 2.133643 over 9775 frames (100.0% kept) avg time waiting for batch 0.021s
2021-01-29 00:11:54,628 INFO [mmi_mbr_train.py:247] batch 40, epoch 0/10 global average objf: 2.321998 over 400056.0 frames (100.0% kept), current batch average objf: 1.984162 over 9862 frames (100.0% kept) avg time waiting for batch 0.017s
```

csukuangfj commented:
Here is a screenshot for epoch 0.  The tensorboard log can be visualized at 
https://tensorboard.dev/experiment/YJq7HAFjSPq0nhZ8vYevhQ/#scalars&_smoothingWeight=0

The training seems working. 

The validation time dominates the total training time: it takes 10 minutes for each validation run. Will reduce the frequency of invoking it.

<img width="1055" alt="Screen Shot 2021-01-29 at 9 30 28 AM" src="https://user-images.githubusercontent.com/5284924/106219618-bcb86080-6214-11eb-993a-fd9c1dffb2a0.png">

pzelasko commented:
BTW The tensor board dev plots will get fixed when you merge master, as I ran into the same issue before. 

> Wiadomość napisana przez Fangjun Kuang <notifications@github.com> w dniu 1/28/21, o godz. 20:35:
> 
> ﻿
> Here is a screenshot for epoch 0. The tensorboard log can be visualized at
> https://tensorboard.dev/experiment/YJq7HAFjSPq0nhZ8vYevhQ/#scalars&_smoothingWeight=0
> 
> The training seems working.
> 
> The validation time dominates the total training time: it takes 10 minutes for each validation run. Will reduce the frequency of invoking it.
> 
> 
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or unsubscribe.

csukuangfj commented:
Thanks. Will merge master into this pull-request once the 10 epochs are finished.

---

WER of epoch 0:

```
2021-01-29 09:59:08,133 INFO [mmi_mbr_decode.py:295] %WER 17.85% [9384 / 52576, 961 ins, 1304 del, 7119 sub ]
```

danpovey commented:
Amazing work!!
BTW we shouldn't put too much hopes on MBR actually improving WER as-is.
It is usually a toss-up between MBR and MMI.
This is a stepping stone to greater possibilities.
Should probably verify that if the sign of the MBR part of the objf is
flipped, things get worse.
I believe in Torch they have automated gradient checks, and maybe a
framework for doing them, IDK if anyone has time to look into that,
to help verify that our code is correct (since it has many pieces).



On Fri, Jan 29, 2021 at 10:06 AM Fangjun Kuang <notifications@github.com>
wrote:

> Thanks. Will merge master into this pull-request once the 10 epochs are
> finished.
> ------------------------------
>
> WER of epoch 0:
>
> 2021-01-29 09:59:08,133 INFO [mmi_mbr_decode.py:295] %WER 17.85% [9384 / 52576, 961 ins, 1304 del, 7119 sub ]
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/k2-fsa/snowfall/pull/84#issuecomment-769523928>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLO73TLXZ2SWRVAC3EZDS4IJZXANCNFSM4WWYVBQA>
> .
>

csukuangfj commented:
> Should probably verify that if the sign of the MBR part of the objf is
flipped, things get worse.

Can it be verified by plotting the MBR loss separately?

danpovey commented:
Yes I think so, the MBR loss should be affected more than the other losses
or the WER.


On Fri, Jan 29, 2021 at 12:13 PM Fangjun Kuang <notifications@github.com>
wrote:

> Should probably verify that if the sign of the MBR part of the objf is
> flipped, things get worse.
>
> Can it be verified by plotting the MBR loss separately?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/k2-fsa/snowfall/pull/84#issuecomment-769562639>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLO7JRQCTUGLKGEDU2DTS4IYVXANCNFSM4WWYVBQA>
> .
>

csukuangfj commented:
@pzelasko I have integrated your fixes from the master branch. Thanks!

---

Below is a screenshot showing how the total loss, MMI loss, and MBR loss change during the training. (I haved renamed objf to loss)

You can see that the MBR loss decreases in a similar way like MMI loss.

The training is still running. The tensorboard log is available at
https://tensorboard.dev/experiment/REra39ZqT4mt60wjz1EX8A/#scalars

![Screen Shot 2021-01-30 at 06 15 20](https://user-images.githubusercontent.com/5284924/106333028-d4e6b900-62c2-11eb-8720-0f6c75705caf.png)

csukuangfj commented:
Below is the WER for each epoch of this pull-request:
```
### epoch-0
2021-01-30 06:08:26,697 INFO [mmi_mbr_decode.py:295] %WER 18.83% [9900 / 52576, 1003 ins, 1318 del, 7579 sub ]

### epoch-1
2021-01-30 10:22:39,902 INFO [mmi_mbr_decode.py:295] %WER 15.19% [7988 / 52576, 935 ins, 884 del, 6169 sub ]

### epoch-2
2021-01-30 16:06:19,820 INFO [mmi_mbr_decode.py:295] %WER 13.45% [7070 / 52576, 867 ins, 785 del, 5418 sub ]

### epoch-3
2021-01-31 06:42:16,209 INFO [mmi_mbr_decode.py:295] %WER 13.19% [6936 / 52576, 866 ins, 709 del, 5361 sub ]

### epoch-4
2021-01-31 06:44:22,864 INFO [mmi_mbr_decode.py:295] %WER 13.31% [7000 / 52576, 971 ins, 645 del, 5384 sub ]

### epoch-5
2021-01-31 14:45:19,317 INFO [mmi_mbr_decode.py:295] %WER 12.69% [6673 / 52576, 795 ins, 701 del, 5177 sub ]

### epoch-6
2021-01-31 15:05:14,329 INFO [mmi_mbr_decode.py:295] %WER 12.75% [6701 / 52576, 876 ins, 618 del, 5207 sub ]

### epoch-7
2021-01-31 20:44:52,752 INFO [mmi_mbr_decode.py:295] %WER 11.96% [6286 / 52576, 826 ins, 632 del, 4828 sub ]

### epoch-8
2021-02-01 06:20:12,109 INFO [mmi_mbr_decode.py:295] %WER 11.91% [6263 / 52576, 808 ins, 650 del, 4805 sub ]

### epoch-9
2021-02-01 08:09:31,706 INFO [mmi_mbr_decode.py:295] %WER 11.04% [5805 / 52576, 802 ins, 535 del, 4468 sub ]
```

csukuangfj commented:
Screenshots of tensorboard are given below:

<img width="1038" alt="Screen Shot 2021-02-01 at 9 11 40 AM" src="https://user-images.githubusercontent.com/5284924/106404395-9ccfa980-646d-11eb-8871-d74fa4c711c3.png">

<img width="1048" alt="Screen Shot 2021-02-01 at 9 11 53 AM" src="https://user-images.githubusercontent.com/5284924/106404404-aa852f00-646d-11eb-9ec6-181cb62e9f36.png">


----

Ready for review.

danpovey commented on egs/librispeech/asr/simple_v1/mmi_mbr_decode.py:
```diff
@@ -0,0 +1,305 @@
+
+# Copyright (c)  2020  Xiaomi Corporation (authors: Daniel Povey, Haowen Qiu)
+# Apache 2.0
+
```
At some point we may want to investigate a more generic decoding script, but this is OK for now..

danpovey commented on egs/librispeech/asr/simple_v1/run.sh:
```diff
@@ -7,7 +7,7 @@
 set -eou pipefail
 
-stage=1
+stage=4
```
change to 1?

csukuangfj replied:
Seems you are reviewing an old commit.

danpovey commented on egs/librispeech/asr/simple_v1/mmi_mbr_train.py:
```diff
@@ -0,0 +1,591 @@
+                                  10.0,
+                                  seqframe_idx_name='seqframe_idx')
+
+    den_lats = k2.intersect_dense(den_graph, dense_fsa_vec, 10.0)
```
As an alternative, we can try using decoding_graph here intead of den_graph (and intersect_dense_pruned).

csukuangfj replied:
So in this case, are there only two graphs: `num_graph` and `decoding_graph` ?

danpovey replied:
Yes.  den_graph and decoding_graph are just different ways of making a generic graph.  It's empirical which one works better.

csukuangfj replied:
Some sequences may return `-inf` for total scores.

Should we discard/ignore such kind of sequences?

danpovey replied:
Yes

csukuangfj replied:
Sequences with tot scores being `-inf` are discarded for MMI, should they be kept for MBR training?

csukuangfj replied:
Otherwise, it is nontrivial to fix the arc indexes while creating the sparse matrices when some FSAs are removed.

danpovey replied:
As I described it and set it up, the MBR training will automatically ignore sequences with score of -inf.

danpovey replied:
But either way.

danpovey replied:
wait.. sorry I was talking about the n-best stuff.  Not sure, do it the way you think is best.

danpovey reviewed (commented):
Looks good generally.. a few comments, but the experiments don't have to be done before  we merge.  Make sure the expected WERs are noted somewhere, if they aren't already.

csukuangfj commented on snowfall/training/mmi_mbr_graph.py:
```diff
@@ -0,0 +1,163 @@
+              shape of the `num_graph`.
+              It is the result of compose(ctc_topo, P)
+
+            - decoding_graph: It is the result of compose(ctc_topo, L, G)
```
Thanks! It is 3x faster.

danpovey commented:
I have profiled this in nvtx, and it seems that more time is
<img width="1188" alt="Screen Shot 2021-02-01 at 4 08 39 PM" src="https://user-images.githubusercontent.com/3298747/106431169-d02f2a00-64a7-11eb-8837-1b0ed350e51c.png">

danpovey commented:
... is spent in graph construction than in actual training.  We should try to move this to GPU.
I think it should be mostly possible.  DeviceIntersect() doesn't support epsilons, but by adding self-loops to the word-level FSA and composing with L, I think we'll get the right thing.  We don't need determinization.  Epsilon removal, we have a device version of that.  For composing the phone-level graph with ctc_topo: again, we need to add epsilon self-loops to the phone-level graph, AddEpsilonSelfLoops().

danpovey commented:
OK to merge this though.

danpovey commented:
... also, when duplicating an FSA via [ fsa] * n, that should be done on GPU not CPU.

csukuangfj commented:
> ... is spent in graph construction than in actual training. We should try to move this to GPU.


> ... also, when duplicating an FSA via [ fsa] * n, that should be done on GPU not CPU.



Let me resolve them first.

danpovey commented:
.. also IDK whether the graph preparation is supposed to be happening in a different thread?

danpovey commented:
It seems to be in the same thread as the objective evaluation.

csukuangfj commented:
> .. also IDK whether the graph preparation is supposed to be happening in a different thread?


I believe it is in the same thread. Only the dataloader, i.e., feature extraction, runs in multiple threads.

csukuangfj commented:
> ... is spent in graph construction than in actual training. We should try to move this to GPU.
I think it should be mostly possible.

I confirm that both intersection of `ctc_topo` and `P` and the intersection of `linear_fsa` and `L_inv` can be run
on GPU with `IntersectDevice` after adding self-loops.

The problem is that there is an optional silence at the begining of each word, which introduces an epsilon transitinon.
To `intersect(ctc_topo_P, num_graph)`, we have to move `ctc_topo_P` and `num_graph` to CPU since we cannot add epsilon self-loops to `num_graph` as it is not epslion free.  

@danpovey  How about removing the optional silence before each word?

danpovey commented:
Can't you just remove the epsilons from num_graph on device before adding epsilon self-loops and doing intersection?
I am thinking maybe you can do all these things by the minibatch (and abandon caching).

pzelasko commented:
> > .. also IDK whether the graph preparation is supposed to be happening in a different thread?
> 
> I believe it is in the same thread. Only the dataloader runs in multiple threads.

I think we could add a “hook”/callback to Lhotse K2 dataset to make it create the graph without introducing a dependency on K2. Does that sound interesting to you?

pzelasko commented:
(Then the graph would be created in a different thread possibly speeding this up)

csukuangfj commented:
> Can't you just remove the epsilons from num_graph on device before adding epsilon self-loops and doing intersection?


Yeah, that's an alternative. Didn't think about it. Thanks!

csukuangfj commented:
> I think we could add a “hook”/callback to Lhotse K2 dataset to make it create the graph without introducing a dependency on K2. Does that sound interesting to you?

That's a good idea. I agree. @danpovey  what do you think?

danpovey commented:
Agreed, it sounds like a good feature.

On Mon, Feb 1, 2021 at 10:17 PM Fangjun Kuang <notifications@github.com>
wrote:

> I think we could add a “hook”/callback to Lhotse K2 dataset to make it
> create the graph without introducing a dependency on K2. Does that sound
> interesting to you?
>
> That's a good idea. I agree. @danpovey <https://github.com/danpovey> what
> do you think?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/k2-fsa/snowfall/pull/84#issuecomment-770887265>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAZFLOZ3EZORQMP4GPXQI7TS42ZLXANCNFSM4WWYVBQA>
> .
>

csukuangfj commented:
Requires https://github.com/k2-fsa/k2/pull/651

Will remove the graph cache and build all graphs in terms of FsaVec to reduce the number of kernel calls.

csukuangfj commented:
Now the graph building process is moved to CUDA.
Previously, it took about 1 minute for 10 batches; now it takes only about 13 seconds for 10 batches.
The training is still running. Will report the results tomorrow.

csukuangfj commented:
## WER of this pull-request (total_loss = mmi_loss + mbr_loss)

```
### epoch 0
2021-02-03 09:12:48,337 INFO [mmi_mbr_decode.py:295] %WER 17.62% [9263 / 52576, 936 ins, 1225 del, 7102 sub ]

### epoch 1
2021-02-03 09:13:45,695 INFO [mmi_mbr_decode.py:295] %WER 15.00% [7888 / 52576, 956 ins, 832 del, 6100 sub ]

### epoch 2
2021-02-03 09:15:19,567 INFO [mmi_mbr_decode.py:295] %WER 14.24% [7487 / 52576, 961 ins, 774 del, 5752 sub ]

### epoch 3
2021-02-03 09:16:07,739 INFO [mmi_mbr_decode.py:295] %WER 13.32% [7002 / 52576, 855 ins, 774 del, 5373 sub ]

### epoch 4
2021-02-03 09:17:01,955 INFO [mmi_mbr_decode.py:295] %WER 14.09% [7409 / 52576, 954 ins, 784 del, 5671 sub ]

### epoch 5
2021-02-03 09:18:36,529 INFO [mmi_mbr_decode.py:295] %WER 13.00% [6834 / 52576, 846 ins, 728 del, 5260 sub ]

### epoch 6
2021-02-03 09:19:42,490 INFO [mmi_mbr_decode.py:295] %WER 13.11% [6894 / 52576, 819 ins, 754 del, 5321 sub ]

### epoch 7
2021-02-03 09:20:41,180 INFO [mmi_mbr_decode.py:295] %WER 12.60% [6627 / 52576, 814 ins, 665 del, 5148 sub ]

### epoch 8
2021-02-03 09:21:35,511 INFO [mmi_mbr_decode.py:295] %WER 12.19% [6408 / 52576, 814 ins, 661 del, 4933 sub ]

### epoch 9
2021-02-03 08:02:51,857 INFO [mmi_mbr_decode.py:295] %WER 11.59% [6096 / 52576, 812 ins, 642 del, 4642 sub ]
```

## WER  (total_loss = mmi_loss + 0.5 * mbr_loss)

```
### epoch 0
2021-02-03 09:23:33,477 INFO [mmi_mbr_decode.py:295] %WER 15.67% [8237 / 52576, 839 ins, 1111 del, 6287 sub ]

### epoch 1
2021-02-03 09:24:26,240 INFO [mmi_mbr_decode.py:295] %WER 13.49% [7090 / 52576, 843 ins, 739 del, 5508 sub ]

### epoch 2
2021-02-03 09:25:23,204 INFO [mmi_mbr_decode.py:295] %WER 12.51% [6577 / 52576, 833 ins, 670 del, 5074 sub ]

### epoch 3
2021-02-03 09:26:33,501 INFO [mmi_mbr_decode.py:295] %WER 12.01% [6313 / 52576, 723 ins, 717 del, 4873 sub ]

### epoch 4
2021-02-03 09:27:23,628 INFO [mmi_mbr_decode.py:295] %WER 12.13% [6378 / 52576, 792 ins, 655 del, 4931 sub ]

### epoch 5
2021-02-03 09:28:17,455 INFO [mmi_mbr_decode.py:295] %WER 11.39% [5986 / 52576, 713 ins, 631 del, 4642 sub ]

### epoch 6
2021-02-03 09:29:07,137 INFO [mmi_mbr_decode.py:295] %WER 11.65% [6123 / 52576, 746 ins, 673 del, 4704 sub ]

### epoch 7
2021-02-03 09:29:56,472 INFO [mmi_mbr_decode.py:295] %WER 11.07% [5818 / 52576, 725 ins, 624 del, 4469 sub ]

### epoch 8
2021-02-03 09:30:44,948 INFO [mmi_mbr_decode.py:295] %WER 10.72% [5635 / 52576, 683 ins, 616 del, 4336 sub ]

### epoch 9
2021-02-03 09:31:43,208 INFO [mmi_mbr_decode.py:295] %WER 11.15% [5862 / 52576, 713 ins, 585 del, 4564 sub ]
```

The WER is lower when we use `total_loss = mmi_loss + 0.5 * mbr_loss`.

csukuangfj commented:
## Loss curves (total_loss = mmi_loss + mbr_loss)

https://tensorboard.dev/experiment/3LNsNJK5RcawmPkpIxlydw/#scalars&_smoothingWeight=0

<img width="1055" alt="Screen Shot 2021-02-03 at 9 42 48 AM" src="https://user-images.githubusercontent.com/5284924/106685865-3fba2c00-6604-11eb-8e22-2590d31d9460.png">

<img width="1061" alt="Screen Shot 2021-02-03 at 9 43 02 AM" src="https://user-images.githubusercontent.com/5284924/106685877-4779d080-6604-11eb-8f1a-010e6bd596d1.png">

## Loss curves (total_loss = mmi_loss + 0.5 * mbr_loss)

https://tensorboard.dev/experiment/fRe7q36vQkeDQ5PcW7oCQw/#scalars&_smoothingWeight=0

<img width="1055" alt="Screen Shot 2021-02-03 at 9 44 32 AM" src="https://user-images.githubusercontent.com/5284924/106685988-83ad3100-6604-11eb-87a3-43d60cf52587.png">

<img width="1062" alt="Screen Shot 2021-02-03 at 9 44 41 AM" src="https://user-images.githubusercontent.com/5284924/106686019-8e67c600-6604-11eb-8357-5fcaad141e07.png">

---

> I just mean I'd expect that the MBR loss would not reach quite as good as value as when the weight is 1.0, and the MMI loss would reach a slightly better value.

The mmi_loss changes as you expected.

- For `total_loss = mmi_loss + mbr_loss`, the final `mmi_loss` is about 0.2279 and `mbr_loss` is 0.109
- For `total_loss = mmi_loss + 0.5 * mbr_loss`, the final `mmi_loss` is about 0.1942 and `mbr_loss` is 0.1043

The `mbr_loss` is slightly lower for the latter case.

csukuangfj commented:
This pull-request requires the latest k2 plus https://github.com/k2-fsa/k2/pull/652

danpovey commented:
Great work!!  You can merge when you are ready..

danpovey commented:
OK.  Well at least changing that weight has more effect on the MMI loss than on the MBR loss, but those loss values don't give me great confidence that the autograd is correct for the MBR loss.  Can you try an experiment where in the 1st epoch it's mmi_loss + mbr_loss, and afterward it's just mbr_loss?   (You can still plot the mmi loss).  I want to see whether it just goes crazy.
I want to develop some automated testing for the autograd, have to think about it..

csukuangfj commented:
Sure.

Working on it.
